apps/settings/js/panels/simpin_dialog/simpin_dialog.js

```diff
@@ -344,48 +345,82 @@ define(function() {
+     * @param {Number} options.error.retryCount
+     * @param {String} options.error.name
      * @access private
      * @return {Boolean} - true means close dialog, others mean not
      */
     _handleCardLockError: function(options) {
+      var error = options.error;
       var lockType = options.lockType;
-      var retryCount = options.retryCount;
+      var retryCount = error.retryCount;
+      var errorName = error.name;
 
       // expected: 'pin', 'fdn', 'puk'
       if (!lockType) {
         // we don't know what's going on here, we have close the dialog.
+        console.error('`handleCardLockError` called without a lockType. ' +
+          'This should never even happen.', error);
         return true;
       }
 
+      switch (errorName) {
+        case 'SimPuk2':
+        case 'IncorrectPassword':
+          return this._handleRetryPassword(lockType, retryCount);
+
+        default:
+          DialogService.alert('genericLockError');
+          console.error('Error of type ' + errorName +
+            ' happened coming from an IccCardLockError event', error);
+          return true;
+      }
+    },
+
+    /**
+     * We will handle all retry cases here to make sure we can change to
+     * the right UI for users to continue.
+     *
+     * @memberOf SimPinDialog
+     * @param {String} lockType
+     * @param {Number} retryCount
+     * @access private
+     * @return {Boolean} - true means close dialog, others mean not
+     */
+    _handleRetryPassword: function(lockType, retryCount) {
       // after three strikes, ask for PUK/PUK2
       if (retryCount <= 0) {
         if (lockType === 'pin') {
-          // we leave this for system app
+          // we leave this for system app, so let's close the dialog
           return true;
         } else if (lockType === 'fdn' || lockType === 'pin2') {
           this._initUI('unlock_puk2');
           this._elements.pukInput.focus();
-        } else { // out of PUK/PUK2: we're doomed
-          // TODO: Shouldn't we show some kind of message here?
+          return false;
+        } else {
+          // out of PUK/PUK2: we're doomed
+          DialogService.alert('genericLockError');
           return true;
         }
-        return false;
-      }
+      } else {
+        // We still have retryCount, let users input values again
+        var msgId = (retryCount > 1) ? 'AttemptMsg3' : 'LastChanceMsg';
+        this._showMessage(lockType + 'ErrorMsg', lockType + msgId, {
+          n: retryCount
+        });
+        this._showRetryCount(retryCount);
 
-      var msgId = (retryCount > 1) ? 'AttemptMsg3' : 'LastChanceMsg';
-      this._showMessage(lockType + 'ErrorMsg', lockType + msgId, {
-        n: retryCount
-      });
-      this._showRetryCount(retryCount);
+        if (lockType === 'pin' || lockType === 'fdn') {
+          this._elements.pinInput.value = '';
+          this._elements.pinInput.focus();
+        } else if (lockType === 'puk') {
```

Comment:
And in this case I believe both `newPinInput.value` and `confirmPinInput.value` should also be cleared.